woocommerce/woocommerce-gutenberg-products-block
# Experiment: Try moving validation to @wordpress/data and move `ValidatedTextInput` to checkout package

This PR is a POC for moving the validation system to a `@wordpress/data` store instead of using context and to move a component that relies on this (`ValidatedTextInput`) to the checkout package. It's not intended to be merged as-is.

I've only implemented one action (`setValidationErrors`) and one selector (`getValidationErrors`) for the new data store.

The changes made here are:
- Add reducer, selectors, and actions for the validation store.
- Move `ValidatedTextInput` to the checkout package and update all references to it.
- Move the `ValidationInputError` to the checkout package and update all references to it.
- Update the `ValidationInputError` component to get the data from the store.
- Export `exposedSelectors` from the validation store to allow third parties to use them.

Throughout this PR I've commented some code out and short-circuited a few things to get it working, these will need to be implemented in the store.

### Testing
If you're using redux dev tools, then you'll need to use a dev version of Gutenberg and add the code from Nadir's P2 (p7H4VZ-3gK-p2) to it.

After that:
- Add an item to your cart and go to the Checkout block.
- Empty the First name field and tab out of it, make sure the error shows. Repeat for other fields that expect input.
- Check the store in redux dev tools and see that the errors are being set in there.

Optionally, create an extension and render this component somewhere on the page.

```ts
/**
 * External dependencies
 */
import { exposedSelectors } from '@woocommerce/blocks-checkout';
import { subscribe } from '@wordpress/data';
import { useEffect, useState } from '@wordpress/element';

export const ExampleComponent = ({ data }) => {

	const [shippingErrors, setShippingErrors] = useState(null);
	useEffect( () => subscribe( () => {
		const nextShippingErrors = exposedSelectors.getValidationError( 'shipping-first_name' );
		setShippingErrors( nextShippingErrors );
	} ), [] );
	return (
		<div className="example-component">
			Shipping first name errors? { shippingErrors?.message }
		</div>
	);
};

```

### Changelog

> Add suggested changelog entry here.
